NeoMathEngine/src/CPU/CpuMathEngineSmallMatricesMultiplyDesc.cpp

```diff
@@ -0,0 +1,169 @@
+/* Copyright © 2023 ABBYY
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <CpuMathEngine.h>
+#include <MemoryHandleInternal.h>
+#include <CpuMathEngineSmallMatricesMultiplyDesc.h>
+
+#ifdef NEOML_USE_MKL
+#if FINE_PLATFORM( FINE_WINDOWS ) || FINE_PLATFORM( FINE_LINUX ) || FINE_PLATFORM( FINE_DARWIN )
+#include <mkl.h>
+#else
+#error Unknown platform
+#endif
+#endif // NEOML_USE_MKL
+
+#define NEOML_MAX(a,b)  ((a)>(b)?(a):(b))
+
+//-------------------------------------------------------------------------------------------------------------------------
+
+namespace NeoML {
+
+inline CSmallMatricesMultiplyDesc* CCpuMathEngine::InitSmallMatricesMultiplyDesc(
+	int firstHeight, int firstWidth, int secondWidth, int secondRowSize, int resultWidth,
+	bool resultAdd, bool trans1, bool trans2 ) const
+{
+	return new CCpuSmallMatricesMultiplyDesc(
+		firstHeight, firstWidth, secondWidth, secondRowSize, resultWidth,
+		resultAdd, trans1, trans2 );
+}
+
+inline bool CCpuMathEngine::SmallMatricesMultiply( const CSmallMatricesMultiplyDesc* desc,
+	const CConstFloatHandle& firstHandle, const CConstFloatHandle& secondHandle, const CFloatHandle& resultHandle ) const
+{
+#if FINE_BIT( FINE_32_BIT )
+	return false;
+#endif // FINE_32_BIT
+
+#ifndef NEOML_USE_MKL
+	ASSERT_EXPR( false );
+#endif // !NEOML_USE_MKL
+
+	return smallMatricesMultiply( desc, GetRaw( firstHandle ), GetRaw( secondHandle ), GetRaw( resultHandle ) );
+}
+
+inline bool CCpuMathEngine::smallMatricesMultiply( const CSmallMatricesMultiplyDesc* desc,
+	const float* first, const float* second, float* result ) const
+{
+#if FINE_BIT( FINE_32_BIT )
+	return false;
+#endif // FINE_32_BIT
+
+#ifndef NEOML_USE_MKL
+	ASSERT_EXPR( false );
+#endif // !NEOML_USE_MKL
+
+	if( desc != nullptr ) {
+		PRESUME_EXPR( dynamic_cast<const CCpuSmallMatricesMultiplyDesc*>( desc ) != nullptr );
+		const CCpuSmallMatricesMultiplyDesc* mulDesk = static_cast<const CCpuSmallMatricesMultiplyDesc*>( desc );
+		if( mulDesk->IsValid() ) {
+			mulDesk->Multiply( first, second, result );
```

Comment:
You can reduce code size (and even speed) you put whole `if( desc != nullptr )` under the
```c++
#if FINE_BIT( FINE_64_BIT ) && defined( NEOML_USE_MKL )
    if( desc != nullptr ) {
        PRESUME_EXPR( dynamic_cast<const CCpuSmallMatricesMultiplyDesc*>( desc ) != nullptr );
        const CCpuSmallMatricesMultiplyDesc* mulDesk = static_cast<const CCpuSmallMatricesMultiplyDesc*>( desc );
        if( mulDesc.MklJitter != nullptr && mulDesc.MklKernel != nullptr ) {
            mulDesc.MklKernel( mulDesc.MklJitter, ( float* )first, ( float* )second, result );
        }
    }
```

As a result you wil remove the definition of `CCpuSmallMatricesMultiplyDesc::Multiply` because it has about 15 lines of code for the sake of  1 short call. And also `desc` will become clearer because it will be descriptor of operation without any math inside of it. Descriptor describes operator, engine performs math based on the descriptions.

The only downside is that you have to
```
    ( void ) desc;
    ( vodi ) first;
```
Mb even it's better to have 2 definitions 
```
#if INE_BIT( FINE_64_BIT ) && defined( NEOML_USE_MKL )
inline bool CCpuMathEngine::smallMatricesMultiply ...
{
   ...
}
#else
inline bool CCpuMathEngine::smallMatricesMultiply( const CSmallMatricesMultiplyDesc*, const float*, const float*, float* ) const
{
    return false;
}
#endif
```
but that's up to your taste...

Reply:
fixed